### PR TITLE
Plugin integration

### PR DESCRIPTION
**Pull request by loqo71la**
Integrates sonarqube plug-in with checks, scanner, and parser module.

### CODE REVIEW

**Comment by loqo71la** [comment]
@ziur @marcocdlv @JimmyRojas please review my changes

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckList.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.fundacionjala.enforce.sonarqube.apex.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+
+/**
+ *
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckList.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+package org.fundacionjala.enforce.sonarqube.apex;
+
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.config.Settings;
+import org.sonar.api.resources.AbstractLanguage;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexMetrics.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+package org.fundacionjala.enforce.sonarqube.apex;
+
+import java.util.Arrays;
+import java.util.List;
+import org.sonar.api.measures.CoreMetrics;
+import org.sonar.api.measures.Metric;
+import org.sonar.api.measures.Metrics;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexMetrics.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexPlugin.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.fundacionjala.enforce.sonarqube.apex.cpd.ApexCpdMapping;
+import org.fundacionjala.enforce.sonarqube.apex.rules.ApexCommonRulesDecorator;
+import org.fundacionjala.enforce.sonarqube.apex.rules.ApexCommonRulesEngine;
+import org.fundacionjala.enforce.sonarqube.apex.rules.ApexRulesDefinition;
+import org.sonar.api.SonarPlugin;
+import org.sonar.api.config.PropertyDefinition;
+import org.sonar.api.resources.Qualifiers;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexPlugin.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexProfile.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.fundacionjala.enforce.sonarqube.apex.checks.CheckList;
+import org.sonar.api.profiles.ProfileDefinition;
+import org.sonar.api.profiles.RulesProfile;
+import org.sonar.api.rules.RuleFinder;
+import org.sonar.api.utils.ValidationMessages;
+import org.sonar.squidbridge.annotations.AnnotationBasedProfileBuilder;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexProfile.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+import org.sonar.squidbridge.SquidAstVisitor;
+import org.sonar.squidbridge.api.CheckMessage;
+import org.sonar.squidbridge.api.SourceCode;
+import org.sonar.squidbridge.api.SourceFile;
+import org.sonar.squidbridge.api.SourceFunction;
+import org.sonar.squidbridge.indexer.QueryByParent;
+import org.sonar.squidbridge.indexer.QueryByType;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.nio.charset.Charset;
+import net.sourceforge.pmd.cpd.Tokenizer;
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+import org.sonar.api.batch.AbstractCpdMapping;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.resources.Language;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+import net.sourceforge.pmd.cpd.SourceCode;
+import net.sourceforge.pmd.cpd.TokenEntry;
+import net.sourceforge.pmd.cpd.Tokenizer;
+import net.sourceforge.pmd.cpd.Tokens;
+import org.fundacionjala.enforce.sonarqube.apex.ApexConfiguration;
+import org.fundacionjala.enforce.sonarqube.apex.lexer.ApexLexer;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesDecorator.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+package org.fundacionjala.enforce.sonarqube.apex.rules;
+
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.batch.rule.CheckFactory;
+import org.sonar.api.component.ResourcePerspectives;
+import org.sonar.squidbridge.commonrules.api.CommonRulesDecorator;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesDecorator.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesEngine.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * THE SOFTWARE.
+ */
+package org.fundacionjala.enforce.sonarqube.apex.rules;
+
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+import org.sonar.squidbridge.commonrules.api.CommonRulesEngine;
+import org.sonar.squidbridge.commonrules.api.CommonRulesRepository;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesEngine.java]: missing doc for the entire class

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ */
+package org.fundacionjala.enforce.sonarqube.apex.rules;
+
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+import org.fundacionjala.enforce.sonarqube.apex.checks.CheckList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.squidbridge.annotations.AnnotationBasedRulesDefinition;
+
+/**
+ *
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: missing doc for the entire class

**Comment by JimmyRojas** [comment]
add doc for all classes please

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv please review my changes, I written documentation in required classes.

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.config.Settings;
+import org.sonar.api.resources.AbstractLanguage;
+
+/**
+ * Defines a Apex language.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java]: Defines a Apex language.

Defines an Apex language.

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexPlugin.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import org.fundacionjala.enforce.sonarqube.apex.rules.ApexCommonRulesEngine;
+import org.fundacionjala.enforce.sonarqube.apex.rules.ApexRulesDefinition;
+
+/**
+ * It's an entry-point to declare its extensions.
+ */
+public class ApexPlugin extends SonarPlugin {
+
+    /**
+     * Classes of the implemented extensions.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexPlugin.java]: obtains/returns a List of ...

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+     */
+    private final Checks<SquidAstVisitor<Grammar>> checks;
+
+    /**
+     * Stores a perspective from resources.
+     */
+    private final ResourcePerspectives resourcePerspectives;
+
+    /**
+     * Stores the all source files to be analyzed.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Stores the all source files to be analyzed.

Stores all source files to be analyzed.

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    @Override
+    public boolean shouldExecuteOnProject(Project project) {
+        FilePredicates predicates = fileSystem.predicates();
+        return fileSystem.hasFiles(predicates.and(
+                predicates.hasType(InputFile.Type.MAIN),
+                predicates.hasLanguage(Apex.KEY)));
+    }
+
+    /**
+     * Analyze files for a given project.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Analyze

Analyzes

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+        this.scanner = ApexAstScanner.create(createConfiguration(), visitors.toArray(new SquidAstVisitor[visitors.size()]));
+        FilePredicates p = fileSystem.predicates();
+        scanner.scanFiles(Lists.newArrayList(fileSystem.files(p.and(p.hasType(InputFile.Type.MAIN), p.hasLanguage(Apex.KEY)))));
+
+        Collection<SourceCode> squidSourceFiles = scanner.getIndex().search(new QueryByType(SourceFile.class));
+        save(squidSourceFiles);
+    }
+
+    /**
+     * Return the simple name of the scanner.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Returns

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+     * Returns the apex configuration.
+     *
+     * @return the configuration.
+     */
+    private ApexConfiguration createConfiguration() {
+        return new ApexConfiguration(fileSystem.encoding());
+    }
+
+    /**
+     * Saves an issues form collection of the source code.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Saves an issues

Saves issues

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+
+            saveFilesComplexityDistribution(inputFile, squidFile);
+            saveFunctionsComplexityDistribution(inputFile, squidFile);
+            saveMeasures(inputFile, squidFile);
+            saveIssues(inputFile, squidFile);
+        });
+    }
+
+    /**
+     * Saves a measures from an input file and source file.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Saves a measures

Saves a measure or Saves measures

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    private void saveFilesComplexityDistribution(InputFile sonarFile, SourceFile squidFile) {
+        RangeDistributionBuilder complexityDistribution = new RangeDistributionBuilder(
+                CoreMetrics.FILE_COMPLEXITY_DISTRIBUTION,
+                FILES_DISTRIB_BOTTOM_LIMITS);
+        complexityDistribution.add(squidFile.getDouble(ApexMetric.COMPLEXITY));
+        context.saveMeasure(sonarFile, buildMeasure(complexityDistribution));
+    }
+
+    /**
+     * Saves an issues form input file and source file.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Saves an issues

Saves issues

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+                        .line(message.getLine())
+                        .message(message.getText(Locale.ENGLISH))
+                        .build();
+                issuable.addIssue(issue);
+            }
+        });
+    }
+
+    /**
+     * Builds and return a measure.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexSquidSensor.java]: Builds and return

Builds and returns

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.charset.Charset;
+
+import net.sourceforge.pmd.cpd.Tokenizer;
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+import org.sonar.api.batch.AbstractCpdMapping;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.resources.Language;
+
+/**
+ * Implements of the CpdMapping extension point.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java]: Implements of the CpdMapping

Implements the CpdMapping

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        List<Token> tokens = lexer.lex(new File(fileName));
+        tokens.forEach(token -> {
+            TokenEntry cpdToken = new TokenEntry(getTokenImage(token), fileName, token.getLine());
+            cpdTokens.add(cpdToken);
+        });
+        cpdTokens.add(TokenEntry.getEOF());
+    }
+
+    /**
+     * Return the Token's value.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java]: Return

Returns

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesDecorator.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.batch.rule.CheckFactory;
+import org.sonar.api.component.ResourcePerspectives;
+import org.sonar.squidbridge.commonrules.api.CommonRulesDecorator;
+
+import org.fundacionjala.enforce.sonarqube.apex.Apex;
+
+/**
+ * Enables that decoration of the language.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexCommonRulesDecorator.java]: Enables that decoration

Enables the decoration

**Comment by JimmyRojas** [comment]
see a few comments please

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv I fixed the required observation. please merge this branch

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckList.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Default constructor.
+     */
+    private CheckList() {
+    }
+
+    /**
+     * Builds and returns the custom checks to create Apex's rules.
+     *
+     * @return
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckList.java]: @return ???

File: sonar-apex-plugin/pom.xml [hunk 25]
```diff
@@ -10,20 +10,44 @@
...
+                <configuration>
+                    <pluginKey>Sonarqube Apex</pluginKey>
+                    <pluginClass>org.fundacionjala.enforce.sonarqube.apex.ApexPlugin</pluginClass>
+                    <pluginName>Apex</pluginName>
+                    <pluginDescription>Apex static code analyzer.</pluginDescription>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
 </project>
```
Comment (JimmyRojas) [inline sonar-apex-plugin/pom.xml]: add the missing new line

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    public Apex(Settings settings) {
+        super(KEY, NAME);
+        this.settings = settings;
+    }
+
+    /**
+     * Returns an array with the key and name of the language. For example ["cls", "apex"].
+     *
+     * @return tha string array.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java]: typo "tha"

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    @Override
+    public String[] getFileSuffixes() {
+        String[] suffixes = filterEmptyStrings(settings.getStringArray(FILE_SUFFIXES_KEY));
+        return suffixes.length == 0 ? DEFAULT_FILE_SUFFIXES : suffixes;
+    }
+
+    /**
+     * Returns an array that filtering the empty string.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java]: I don't understand this statement

"Returns an array that filtering the empty string."

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @param stringArray string array to be filtered.
+     * @return the string array.
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/Apex.java]: improve the return statement

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public ApexCpdMapping(Apex language, FileSystem fileSystem) {
+        this.language = language;
+        this.charset = fileSystem.encoding();
+    }
+
+    /**
+     * Builds and returns an {@link ApexTokenizer}.
+     *
+     * @return
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexCpdMapping.java]: @return ???

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Stores an apex configuration.
+     */
+    private final ApexConfiguration config;
+
+    /**
+     * Default constructor to initialize the configuration.
+     *
+     * @param charset
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/cpd/ApexTokenizer.java]: @param charset  ... ??

**Comment by JimmyRojas** [comment]
see more comments, and please be careful before to request a codeReview..

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv apparently everything is fine now, please review my changes

**Comment by JimmyRojas** [comment]
Looks good to me.., thank you

**Comment by marcocdlv** [comment]
It looks good to me too

**Pull request merged**